src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/RazorCompletionItemResolver.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Tooltip;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor.Completion;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text.Adornments;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    internal class RazorCompletionItemResolver : CompletionItemResolver
+    {
+        private readonly LSPTagHelperTooltipFactory _lspTagHelperTooltipFactory;
+        private readonly VSLSPTagHelperTooltipFactory _vsLspTagHelperTooltipFactory;
+
+        public RazorCompletionItemResolver(
+            LSPTagHelperTooltipFactory lspTagHelperTooltipFactory,
+            VSLSPTagHelperTooltipFactory vsLspTagHelperTooltipFactory)
+        {
+            _lspTagHelperTooltipFactory = lspTagHelperTooltipFactory;
+            _vsLspTagHelperTooltipFactory = vsLspTagHelperTooltipFactory;
+        }
+
+        public override Task<VSInternalCompletionItem?> ResolveAsync(
+            VSInternalCompletionItem completionItem,
+            VSInternalCompletionList containingCompletionList,
+            object? originalRequestContext,
+            VSInternalClientCapabilities? clientCapabilities,
+            CancellationToken cancellationToken)
+        {
+            if (originalRequestContext is not IReadOnlyList<RazorCompletionItem> razorCompletionItems)
+            {
+                // Can't recognize the original request context, bail.
+                return Task.FromResult<VSInternalCompletionItem?>(null);
+            }
+
+            var associatedRazorCompletion = razorCompletionItems.FirstOrDefault(completion => string.Equals(completion.DisplayText, completionItem.Label, StringComparison.Ordinal));
+            if (associatedRazorCompletion is null)
+            {
+                Debug.Fail("We should never not be able to find an associated Razor completion item when resolving completions.");
+                return Task.FromResult<VSInternalCompletionItem?>(null);
+            }
+
+            // If the client is VS, also fill in the Description property.
+            var useDescriptionProperty = clientCapabilities?.SupportsVisualStudioExtensions ?? false;
```

Comment:
Everything from here down was a direct copy & paste from the old resolution endpoint